gator/fragments.py

```diff
@@ -123,8 +123,22 @@ def count_specified_regex(contents, regex):
     # finds regex matches, returns their count
     if not is_valid_regex(regex):
         return 0
-    matches = re.findall(regex, contents)
-    num_matches = len(matches)
+
+    new_contents = str(contents)
+    is_match = True
+    num_matches = 0
+    position_start = 0
+    position_end = len(new_contents)
+
+    while is_match:
+        match = re.search(regex, repr(new_contents[position_start:position_end]))
+        if match is not None:
+            num_matches += 1
+            position_start = match.start() + 1
+            position_end = match.end() - 1
+        else:
+            is_match = False
```

Comment:
Why was this changed? It seems the last version was less complex and still should achieve the result.

Reply:
We added this in so that it can iterate through multiple regex. As it was, if there were multiple regex, such as multiple \begins and \ends, it could not get the interior ones thus resulting in not being able to capture more than one of the regex

Reply:
Um... unless I'm misunderstanding, that's a failure of the regex, not of the algorithm. I'm not sure I understand though.

Reply:
For example, in LaTeX, a potential regex could be:
\begin{enumerate}
    hello there
\end{enumerate}
\begin{enumerate}
    hello world
\end{enumerate}

Previously, if you want to check the amount of begin and ends, it would only find the first \begin and the last \end, resulting in only 1 regex for \begin{enumerate} and \end{enumerate} found. Now with these changes, it finds each instance, so it finds both \begin{enumerate} and \end{enumerates}, resulting in the correct 2 instances for the regex.

I hope this explanation helps. If not, let me know.

Reply:
Yes, that is a problem with the regex. You need to use lazy qualifiers ( `?` ) to ensure that the **first** instance of the characters following a `.*` is picked. You should not modify the algorithm for that -- regexes will now not be matched correctly.

Reply:
I see what you mean, if we specify the regex as `\\begin([^;]*?)\\end` it will distinguish multiple groups of regex, like in Simon's example. 
The other problem is with regex examples embedded inside another regex. For example:
\begin{document}
hello
\begin{enumerate}
\item hello
\end{enumerate}
\end{document}

The algorithm doesn't see to distinguish between the first and second `\end`. Not sure if this is a fully unrelated problem but I feel like it applies to the kind of checks that would be performed on an actual LaTeX file.

Reply:
I'm not sure what you mean by `regex` here -- `\\begin([^;]*?)\\end` is a regex, the text is not.
The regex shouldn't distinguish between the two `\end`s -- there is nothing in the regex to do that! A positive lookahead with a backreference to a capturing group for the type of environment would work here. Perhaps brushing up on regular expressions using a resource like [regex 101](https://regex101.com/) would help.

As a side note, the regex you gave doesn't really make sense to detect begin-end blocks; the contents are simply everything that isn't `;`, however that could easily be part of the contents in LaTeX.